hxpipe.1
.TH "HXPIPE" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"

.de d \" begin display
.sp
.in +4
.nf
.ft CR
.CDS
..
.de e \" end display
.CDE
.in -4
.fi
.ft R
.sp
..

.SH NAME
hxpipe \- convert XML file to a format easier to parse with Perl or AWK
.SH SYNOPSIS
.B hxpipe
.RB "[\| " \-l " \|]"
.RB "[\| " \-\- " \|]"
.RI "[\| " file-or-URL " \|]"
.SH DESCRIPTION
.B hxpipe
parses an HTML or XML file and outputs a line-oriented representation
of it that is well suited to further processing with AWK or similar
tools. The format is similar to the ESIS (Element Structure
Information Set) that is output by nsgmls/onsgmls.
.LP
The reverse operation, converting back to mark-up, is performed by the
.B hxunpipe
program.
.LP
The output format is as follows:
.TP 10
<!--comment-->
Comments are output as
.d
*comment
.e
I.e., a single line starting with "*" followed by the text of the
comment. Line feeds, carriage returns and tabs in the text are written
as "\\n", "\\r" and "\\t", respectively. Text that looks like a
numerical character entity is written with the "&" replaced by "\\".
The line ends with a line feed.
.IP ""
Note that onsgmls outputs comments starting with a "_" instead of a
"*" and doesn't replace the "&" of numerical character entities by
"\\" (and by default it omits comments altogether).
.TP
<?processing instruction>
Processing instructions are output as
.d
?processing instruction
.e
I.e., a single line starting with a "?" followed by the text of the
processing instruction. The text is escaped as for comments (see
above).
.TP
<!DOCTYPE root PUBLIC "-//foo//DTD bar//EN" "http://example.org/dtd">
DOCTYPEs are output as one of the following:
.d
!root "-//foo//DTD bar//EN" http://example.org/dtd
!root "-//foo//DTD bar//EN"
!root "" http://example.org/dtd
!root ""
.e
for respectively: a DOCTYPE with (1) both a public and a system
identifier, (2) only a public identifier, (3) only a system
identifier, or (4) neither of the two. I.e., a single line starting
with a "!", followed by a space and a possibly empty quoted string,
followed optionally by a space and arbitrary text. Note the quotes for
the public identifier and the absence of quotes for the system
identifier.
.TP
<elt att1="value1" att2="value2">
A start tag is output as
.d
Aatt1 CDATA value1
Aatt2 CDATA value2
(elt
.e
I.e., as zero or more lines for the attributes and one line for the
element type. Each line for an attribute starts with "A" followed by
the name of the attribute, a space, the literal string "CDATA",
another space, and the attribute value. The text of the attribute
value is escaped as for comments (see above). The line for the element
type starts with "(" followed by the element type.
.IP ""
.B hxpipe
does not read DTDs and assumes that attributes are always CDATA. It
never generates other types (IMPLIED, TOKEN, ID, etc.), unlike
onsgmls.
.TP
</elt>
End tags are output as
.d
)elt
.e
I.e., as a line starting with ")" followed by the element type.
.TP
<empty att1="val1" att2="val2"/>
Empty elements (in XML) are output as
.d
Aatt1 CDATA val1
Aatt2 CDATA val2
|empty
.e
I.e., as zero or more lines for attributes and one line starting with
"|" followed by the element type.
.IP ""
Note that
.B onsgmls
never outputs "|". (However, it can optionally output a line
consisting of a single "e" just before the "(" line, to indicate that
the element is empty.)
.TP
text
Text is output as
.d
\-text
.e
I.e., as a single line starting with a "\-". The text is escaped as
for comments (see above).
.TP
line numbers
When the
.B \-l
option is in effect,
.B hxpipe
will intersperse the output with lines of the form
.d
L12
.e
where "12" is replaced with the line number in the source where the
next output came from.
.LP
.B hxpipe
does not normalize the input and does not add mising tags. It is thus
possible that there are unequal numbers of "(" and ")" lines. If it is
important that every start tag is matched by an end tag, pipe the
input through
.B hxnormalize -x
first.
.SH OPTIONS
The following options are supported:
.TP 10
.B \-l
Add "L" lines to the output to indicate the line numbers in the
source.
.SH OPERANDS
The following operand is supported:
.TP 10
.I file-or-URL
The name or URL of an HTML file. If absent, standard input is read
instead.
.SH "EXIT STATUS"
The following exit values are returned:
.TP 10
.B 0
Successful completion.
.TP
.B > 0
An error occurred in the parsing of the HTML file.
.B hxpipe
will try to correct the error and produce output anyway.
.SH ENVIRONMENT
To use a proxy to retrieve remote files, set the environment variables
.B http_proxy
and
.BR ftp_proxy "."
E.g.,
.B http_proxy="http://localhost:8080/"
.SH BUGS
.LP
The error recovery for incorrect HTML is primitive.
.B hxnormalize
can currently only retrieve remote files over HTTP. It doesn't handle
password-protected files, nor files whose content depends on HTTP
"cookies."
.SH "SEE ALSO"
.BR hxunpipe (1),
.BR onsgmls (1).


ChangeLog
2019-10-05  Bert Bos  <bert@w3.org>

	* Published version 7.8.

	* textwrap.c, langinfo.c, hxnormalize.c: Added knowledge about
	languages that do not use spaces between words. In such languages,
	a newline should not be converted to a space in outc() in
	textwrap.c, but only to a break opportunity.

2019-08-28  Bert Bos  <bert@w3.org>

	* hxtoc.c: The element to group headings in HTML5 is called
	HGROUP, not HEADER. The heading of a section (SECTION, ARTICLE,
	etc.) need not be the first element, there may be non-header
	elements before it.

	* hxwls.c: Print "longdesc", "classid" or "codebase" in the second
	column for the corresponding attribute. Also recognize srcset
	(somewhat).

2018-08-03  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Added option -X to indicate the input is XML
	instead of HTML. Handle conversion of CDATA elements to XML by
	escaping < and & instead of adding <![CDATA[. Added corresponding
	test normalize13.sh.

2018-04-29  Bert Bos  <bert@w3.org>

	* Released version 7.7.

	* dtd.hash: Don't include the arguments in the forward declaration
	of lookup_element(), because those arguments differ slightly
	depending on which version of gperf is used to generate dtd.c:
	"unsigned int" in gperf 3.0 vs "size_t" in gperf 3.1.

	* conficgure.ac: Check for libiconv with the AM_ICONV macro (from
	gettext) instead of AC_SEARCH_LIBS. (Advice from Christian
	Weisgerber, received via Anthony Bentley). That in turn requires
	adding config.rpath to the EXTRA_DIST in Makefile.am.

2018-02-23  Bert Bos  <bert@w3.org>

	* tests/index10.sh: Added a test for a term database.

	* hxindex.c: Replaced fgets() by getline(), to allow lines of
	arbitrary length in the term database.

2018-02-15  Bert Bos  <bert@w3.org>

	* tests/index9.sh: Added a test for hxindex's option -N.

	* tests/index5.sh: Updated for hxindex's option -N.

	* tests/cite5.sh: Added the example from hxmkbib.1 as a test.

	* tests/normalize12.sh: Added a test for the NOSCRIPT element.

	* hxindex.1: Added option -N.

	* hxmkbib.1: The example was missing some "%" signs.

	* hxmkdir.c: Read the template before reading the auxfile, so that
	a pipe like "hxcite -a auxfile bib template | hxmkbib -a auxfile
	bib" becomes possible. hxcite will have the time to finishe
	writing the auxfile before hxmkbib reads it.

2018-02-12  Bert Bos  <bert@w3.org>

	* hxindex.c: Added option -N.

	* dtd.hash: Since HTML5, the NOSCRIPT element is phrase-level
	instead of block-level.

2018-02-09  Bert Bos  <bert@w3.org>

	* Version 7.6 released.

	* selmatch.c: matches_sel() could read uninitialized memory
	(h->parent where h is the root of a tree) potentially leading to a
	SIGBUS. Thanks to Anthony J. Bentley for finding this bug.

2018-01-20  Bert Bos  <bert@w3.org>

	* Version 7.5 released.

	* hxcite.1: Spelling error.

	* hxincl.1: Spelling error.

	* url.c: Fixed unsafe way to call warnx(). Thanks to Robin
	Naundorf for these three corrections.

2017-12-08  Bert Bos  <bert@w3.org>

	* tests/normalize11.sh: Added to test for "main".

	* dtd.hash: Added the HTML5 element "main".

2017-12-07  Bert Bos  <bert@w3.org>

	* hxwls.c: Added support for HTML5 elements (source, audio, video)
	and for the classid and codebase attribute. Fixed an error: input
	has a src attribute, not href.

2017-11-24  Bert Bos  <bert@w3.org>

	* Version 7.4 released.

	* hxremove.1, hxselect.1: Updated.

	* tests/remove4.sh, tests/select18.sh, tests/select19.sh,
	tests/select20.sh, tests/select21.sh, tests/select22.sh,
	tests/remove5.sh tests/remove6.sh tests/remove7.sh,
	tests/select23.sh tests/select24.sh: Added.

	* selmatch.c: Now handles the ':not()' pseudo-class and
	comma-separated selectors. Fixed a bug in matching the ':root'
	pseudo-class.

	* hxremove.c, hxselect.c: parse_selector() has different
	arguments.

	* selector.c: Now parses the ':not()' pseudo-class and
	comma-separated selectors. Errors are now printed with a newline
	at the end. Added new functions to dump (serialize) a selector to
	a file, useful for debugging.

	* openurl.c: Functions now either return CURLcode or CURLMcode,
	but not some mixture.

2017-11-11  Bert Bos  <bert@w3.org>

	* dtd.hash: Added some HTML5 elements: source, track, wbr (because
	they are empty) and audio and video (as parents for source and
	track).

	* toc4.sh, normalize10.sh: Added tests for the <source> element.

	* Released version 7.3.

2017-10-16  Bert Bos  <bert@w3.org>

	* normalize9.sh: Added test for A inside FIGCAPTION.

	* dtd.hash: An A can now be inside a FIGCAPTION. Made source
	shorter and more readable with some macros.

2017-10-06  Bert Bos  <bert@w3.org>

	* Released version 7.2.

	* tests/extract1.sh: Added option -N to nc, to close the
	connection after EOF on stdin, because hxextract (with libcurl)
	seems to wait for the closing of the connection.

	* tests/unent1.ssh, tests/unent2.ssh, tests/unent3.ssh,
	tests/unent4.ssh, tests/unent5.ssh: Added tests for hxunent.

	* hxunentmain.c: Fixed bug reported by Laurent Martelli: When a
	line was longer than the 4096-byte buffer, an entity could be on
	the edge between two buffers and not be recognized.

2017-08-13  Bert Bos  <bert@w3.org>

	* extract1.sh: return code 77 (skip) rather than 0 when netcat is
	unavailable.

	* tests/pipe4.sh, tests/select01.sh, ..., tests/select17.sh,
	tests/unpipe6.sh: New tests.

	* scan.l: Accept non-ASCII characters in names (element names,
	attribute names, etc.)

	* hxselect.c, hxremove.c: Put their common code in
	selmatch.c. hxselect now uses the data structures from tree.c,
	just like hxremove.c. The non-element selector '::attr(...)' is
	now also supported.

	* selector.c, selmatch.c: Added support for more structural
	pseudo-classes (last-child, nth-last-of-type, etc.). Generalized
	support for backslash escapes. Support hexadecimal backslash
	escapes also outside ASCII range (such characters are converted to
	UTF-8). Bug fixes.

2016-10-17  Bert Bos  <bert@w3.org>

	* Released version 7.1.


hxpipe.c
/*
 * pipe - output HTML/XML in canonical ("sgmls" form
 *
 * Parse HTML/XML and output in approximate "nsgmls" format. Some of
 * the differences are that comments are also printed (see * below),
 * that implied attributes are not, and that entities are left
 * unexpanded. Use "unent" to expand entities to UTF-8.
 *
 * The program doesn't interpret the source in any way, and doesn't
 * read DTDs. That means that, e.g., end tags are not automatically
 * added. Pipe the source through normalize(1) first in order to
 * convert HTML to XML and infer missing tags.
 *
 * The possible command characters and arguments are as follows:
 *
 *     (gi
 *
 *	    The start of an element whose generic identifier is gi.
 *	    Any attributes for this element will have been speci- fied
 *	    with A commands.
 *
 *     )gi
 *
 *	    The end of an element whose generic identifier is gi.
 *
 *     |gi
 *
 *	    An empty element (an element whose tag in the source ended
 *	    with a slash). Any attributes will have been specified
 *	    with A commands. (Note that this distinguishes empty
 *	    elements from elements that happen to have no content,
 *	    even though XML doesn't.)
 *
 *     -data
 *
 *	    Data.
 *
 *     ?pi
 *
 *	    A processing instruction with data pi.
 *
 *     *comment
 *
 *	    A comment
 *
 *     Aname type val
 *
 *	    The next element to start has an attribute name with value
 *	    val and type type. Implied attribute are not shown. All
 *	    attributes are assumed to be of type CDATA, because pipe
 *	    doesn't read DTDs. The exceptions are "xml:id" and
 *	    "xmlid", which are assumed to be of type TOKEN.
 *
 *     !root "fpi" url
 *     !root "fpi"
 *     !root "" url
 *
 *	    A document type declaration. The fpi (public identifier)
 *	    is a quoted string. If there is no fpi, the string is
 *	    empty: "". If there is no url, itis omitted.
 *
 *     Llineno
 *
 *	    Set the current line number. This will be output only if
 *	    the -l option has been given.
 *
 * Part of HTML-XML-utils, see:
 * http://www.w3.org/Tools/HTML-XML-utils/
 *
 * Copyright © 1994-2012 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/copyright-software
 *
 * Author: Bert Bos
 * Created: 2 Dec 1998
 * Version: $Id: hxpipe.c,v 1.9 2017/11/24 09:50:25 bbos Exp $
 *
 **/
#include "config.h"
#include <stdio.h>
#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
#include <ctype.h>
#if STDC_HEADERS
# include <string.h>
#else
# ifndef HAVE_STRCHR
#  define strchr index
#  define strrchr rindex
# endif
#endif
#include <stdlib.h>
#include <assert.h>
#include <stdbool.h>
#include "export.h"
#include "types.e"
#include "html.e"
#include "scan.e"
#include "dict.e"
#include "openurl.e"
#include "errexit.e"

#define XMLID "{http://www.w3.org/XML/1998/namespace}id"

static bool has_error = false;
static bool in_text = false;
static bool linenumbering = false;


/* escape -- print a string with certain characters escaped */
static void escape(const string t)
{
  string s;

  for (s = t; *s; s++)
    switch (*s) {
      case '\r': printf("\\r"); break;
      case '\t': printf("\\t"); break;
      case '\n': printf("\\n"); break;
      case '\\': printf("\\\\"); break;
      case '&': if (*(s+1) == '#') printf("\\"); else printf("&"); break;
      default: putchar(*s);
    }
}


/* --------------- implements interface api.h -------------------------- */

/* handle_error -- called when a parse error occurred */
void handle_error(void *clientdata, const string s, int lineno)
{
  fprintf(stderr, "%d: %s\n", lineno, s);
  has_error = true;
}

/* start -- called before the first event is reported */
void* start(void)
{
  return NULL;
}
  
/* end -- called after the last event is reported */
void end(void *clientdata)
{
  if (in_text) {putchar('\n'); in_text = false;}
}

/* handle_comment -- called after a comment is parsed */
void handle_comment(void *clientdata, string commenttext)
{
  if (in_text) {putchar('\n'); in_text = false;}
  if (linenumbering) printf("L%d\n", lineno);
  putchar('*');
  escape(commenttext);
  putchar('\n');
}

/* handle_text -- called after a text chunk is parsed */
void handle_text(void *clientdata, string text)
{
  /* There may be several consecutive calls to this routine. The
   * variable 'in_text' is used to put the text of all of them on the
   * same line.
   **/
  if (! in_text) {
    if (linenumbering) printf("L%d\n", lineno);
    putchar('-');
    in_text = true;
  }
  escape(text);
}

/* handle_decl -- called after a declaration is parsed */
void handle_decl(void *clientdata, string gi, string fpi,
		 string url)
{
  if (in_text) {putchar('\n'); in_text = false;}
  if (linenumbering) printf("L%d\n", lineno);
  printf("!%s \"%s\" %s\n", gi, fpi ? fpi : "", url ? url : "");
}

/* handle_pi -- called after a PI is parsed */
void handle_pi(void *clientdata, string pi_text)
{
  if (in_text) {putchar('\n'); in_text = false;}
  if (linenumbering) printf("L%d\n", lineno);
  putchar('?');
  escape(pi_text);
  putchar('\n');
}

/* print_attrs -- print attributes */
void print_attrs(const pairlist attribs)
{
  pairlist p;

  for (p = attribs; p; p = p->next) {
    putchar('A');
    printf("%s", p->name);
    if (eq(p->name, "xmlid") || eq(p->name, "xml:id") ||
	eq(p->name, XMLID)) printf(" TOKEN ");
    else printf(" CDATA ");
    if (p->value) escape(p->value); else printf("%s", p->name);
    putchar('\n');
  }
}

/* handle_starttag -- called after a start tag is parsed */
void handle_starttag(void *clientdata, string name, pairlist attribs)
{
  if (in_text) {putchar('\n'); in_text = false;}
  print_attrs(attribs);
  if (linenumbering) printf("L%d\n", lineno);
  putchar('(');
  printf("%s", name);
  putchar('\n');
}

/* handle_emptytag -- called after an empty tag is parsed */
void handle_emptytag(void *clientdata, string name, pairlist attribs)
{
  if (in_text) {putchar('\n'); in_text = false;}
  print_attrs(attribs);
  if (linenumbering) printf("L%d\n", lineno);
  putchar('|');
  printf("%s", name);
  putchar('\n');
}

/* handle_endtag -- called after an endtag is parsed (name may be "") */
void handle_endtag(void *clientdata, string name)
{
  if (in_text) {putchar('\n'); in_text = false;}
  if (linenumbering) printf("L%d\n", lineno);
  putchar(')');
  printf("%s", name);
  putchar('\n');
}

/* --------------------------------------------------------------------- */

/* usage -- print usage message and exit */
static void usage(string prog)
{
  fprintf(stderr, "Usage: %s [-l] [-v] [html-file-or-url]\n", prog);
  exit(2);
}

int main(int argc, char *argv[])
{
  int c, status = 200;

  /* Bind the parser callback routines to our handlers */
  set_error_handler(handle_error);
  set_start_handler(start);
  set_end_handler(end);
  set_comment_handler(handle_comment);
  set_text_handler(handle_text);
  set_decl_handler(handle_decl);
  set_pi_handler(handle_pi);
  set_starttag_handler(handle_starttag);
  set_emptytag_handler(handle_emptytag);
  set_endtag_handler(handle_endtag);

  /* Parse command line arguments */
  while ((c = getopt(argc, argv, "lv")) != -1)
    switch (c) {
    case 'l': linenumbering = true; break;
    case 'v': printf("Version: %s %s\n", PACKAGE, VERSION); return 0;
    case '?': usage(argv[0]); break;
    default: assert(!"Cannot happen");
    }

  if (optind == argc) yyin = stdin;
  else if (optind == argc - 1 && eq(argv[optind], "-")) yyin = stdin;
  else if (optind == argc - 1) yyin = fopenurl(argv[optind], "r", &status);
  else usage(argv[0]);

  if (yyin == NULL) {perror(argv[optind]); exit(1);}
  if (status != 200) errexit("%s : %s\n", argv[optind], http_strerror(status));

  if (yyparse() != 0) exit(3);

  return has_error ? 1 : 0;
}
